# Evaluación 1
Las datas frames elegidas corresponden a las siguientes páginas web: 
1. Population by country: 
2. Green House Emissions by country: 
3. Freshwater withdrawals by country: 

In [203]:
#Links to databases in folder "data" in repository "Evalucion1"
population_Link="https://www.cia.gov/the-world-factbook/field/population/country-comparison/"
h2O_withdrawl_Link="https://github.com/SpatialDataScience23-2/Evaluacion1/raw/main/data/freshwater_withdrawals.csv"
pm2_5_Link="https://github.com/SpatialDataScience23-2/Evaluacion1/raw/main/data/pm2_5.csv"

import pandas as pd
population=pd.read_html(population_Link, header=0, flavor= 'bs4')
h2O_withdrawl=pd.read_csv(h2O_withdrawl_Link)
pm2_5=pd.read_csv(pm2_5_Link)

population_new=population[0]

Rank                  Country  Unnamed: 2 Date of Information
0       1                    China  1413142846           2023 est.
1       2                    India  1399179585           2023 est.
2       3            United States   339665118           2023 est.
3       4                Indonesia   279476346           2023 est.
4       5                 Pakistan   247653551           2023 est.
..    ...                      ...         ...                 ...
232   233                  Tokelau        1647           2019 est.
233   234          Paracel Islands        1440      July 2014 est.
234   235  Holy See (Vatican City)        1000           2022 est.
235   236  Cocos (Keeling) Islands         596      July 2014 est.
236   237         Pitcairn Islands          50           2021 est.

[237 rows x 4 columns]

### df1: h2O_withdrawl
1. In this data base we have the information of the total annual freswater withdrawls by each country from 1975-2019
2. That's alot of data, so I decided to restrict it to just the variable values of year 2018
3. We recall the reduced data base with the name "h2O"

In [53]:
h2O_new=h2O_withdrawl.copy()
h2O_new.Year.value_counts()

Year
2018    195
2019    195
2017    195
2016    195
2015    195
2014    195
2013    181
2012    181
2011    177
2010    177
2009    176
2008    176
2007    173
2006    171
2005    167
2004    165
2003    164
2002    163
2001    160
2000    158
1999    148
1998    147
1997    146
1996    145
1995    143
1994    137
1993    131
1992    122
1991    110
1990    104
1989     91
1988     90
1987     89
1986     62
1985     61
1984     56
1983     55
1982     55
1981     54
1980     54
1976     35
1979     35
1978     35
1977     35
1975     34
1974     17
1970     16
1971     16
1972     16
1973     16
1965      2
1966      2
1967      2
1968      2
1969      2
1962      1
1963      1
1964      1
Name: count, dtype: int64

In [204]:
#we restrict the Year value to only 2018
h2O_new_1=h2O_new[h2O_new["Year"]==2018]
h2O_new_1.Year.value_counts()
#we eliminate the column "Year"
h2O=h2O_new_1.drop(columns="Year")
h2O

Entity      Code  \
43            Afghanistan       AFG   
73                Albania       ALB   
123               Algeria       DZA   
156                Angola       AGO   
186   Antigua and Barbuda       ATG   
...                   ...       ...   
5728   West Bank and Gaza       NaN   
5734                World  OWID_WRL   
5764                Yemen       YEM   
5792               Zambia       ZMB   
5825             Zimbabwe       ZWE   

      Annual freshwater withdrawals, total (billion cubic meters)  
43                                         2.028200e+10            
73                                         9.580000e+08            
123                                        9.802000e+09            
156                                        7.057000e+08            
186                                        4.400000e+06            
...                                                 ...            
5728                                       2.997000e+08            
5734                                       3.955609e+12            
5764                                       3.565000e+09            
5792                                       1.572000e+09            
5825                                       3.771380e+09            

[195 rows x 3 columns]

### df2: PM2.5
1. In this data base we have the information of the PM2.5 by each country
2. There are value for 2018 and 2019, so I decided to restrict it to just the variable values of year 2018
3. We recall the reduced data base with the name "pm25"

In [205]:
pm2_5.Year.value_counts()

Year
2018    239
2019    239
Name: count, dtype: int64

In [206]:
pm2_5_new=pm2_5.copy()
pm2_5_new_1=pm2_5_new[pm2_5_new["Year"]==2018]
pm2_5_new_1.YEA.value_counts()

YEA
2018    239
Name: count, dtype: int64

In [207]:
pm25=pm2_5_new_1.copy()
byeColumns=['Year','YEA']
pm25.drop(columns=byeColumns,inplace=True)
pm25

COU                                          Country  \
0          AUS                                        Australia   
2          AUT                                          Austria   
4          BEL                                          Belgium   
6          CAN                                           Canada   
8          CZE                                   Czech Republic   
..         ...                                              ...   
468       OECD                                     OECD - Total   
470      OECDE                                    OECD - Europe   
472       EA19                         Euro area (19 countries)   
474  EU27_2020  European Union – 27 countries (from 01/02/2020)   
476        MYT                                          Mayotte   

     SMALL_SUBNATIONAL_REGION Small subnational region  \
0                         NaN           Not applicable   
2                         NaN           Not applicable   
4                         NaN           Not applicable   
6                         NaN           Not applicable   
8                         NaN           Not applicable   
..                        ...                      ...   
468                       NaN           Not applicable   
470                       NaN           Not applicable   
472                       NaN           Not applicable   
474                       NaN           Not applicable   
476                       NaN           Not applicable   

    LARGE_SUBNATIONAL_REGION Large subnational region     VAR  \
0                      TOTAL                    Total  PWM_EX   
2                      TOTAL                    Total  PWM_EX   
4                      TOTAL                    Total  PWM_EX   
6                      TOTAL                    Total  PWM_EX   
8                      TOTAL                    Total  PWM_EX   
..                       ...                      ...     ...   
468                    TOTAL                    Total  PWM_EX   
470                    TOTAL                    Total  PWM_EX   
472                    TOTAL                    Total  PWM_EX   
474                    TOTAL                    Total  PWM_EX   
476                    TOTAL                    Total  PWM_EX   

                              Variable Unit Code                        Unit  \
0    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
2    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
4    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
6    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
8    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
..                                 ...       ...                         ...   
468  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
470  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
472  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
474  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
476  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   

     PowerCode Code PowerCode  Reference Period Code  Reference Period  \
0                 0     Units                    NaN               NaN   
2                 0     Units                    NaN               NaN   
4                 0     Units                    NaN               NaN   
6                 0     Units                    NaN               NaN   
8                 0     Units                    NaN               NaN   
..              ...       ...                    ...               ...   
468               0     Units                    NaN               NaN   
470               0     Units                    NaN               NaN   
472               0     Units                    NaN               NaN   
474               0     Uni

### df3: population
1. In this data base we have the information of the total population by each country
2. There is data from year 2000 to 2023
3. I will restrict it to the population value for year 2018
4. The filtered df will have the name "pop"

In [209]:
population_new
population_new.columns

Index(['Rank', 'Country', 'Unnamed: 2', 'Date of Information'], dtype='object')

In [213]:
population_new.columns=population_new.columns.str.replace(' ','_')
population_new.rename(columns={'Unnamed:_2':'Population'},inplace=True)
population_new

Rank                  Country  Population Date_of_Information
0       1                    China  1413142846           2023 est.
1       2                    India  1399179585           2023 est.
2       3            United States   339665118           2023 est.
3       4                Indonesia   279476346           2023 est.
4       5                 Pakistan   247653551           2023 est.
..    ...                      ...         ...                 ...
232   233                  Tokelau        1647           2019 est.
233   234          Paracel Islands        1440      July 2014 est.
234   235  Holy See (Vatican City)        1000           2022 est.
235   236  Cocos (Keeling) Islands         596      July 2014 est.
236   237         Pitcairn Islands          50           2021 est.

[237 rows x 4 columns]

In [215]:
h2O.shape, population_new.shape, pm25.shape

((195, 3), (237, 4), (239, 17))

In [219]:
pm25.columns

Index(['COU', 'Country', 'SMALL_SUBNATIONAL_REGION',
       'Small subnational region', 'LARGE_SUBNATIONAL_REGION',
       'Large subnational region', 'VAR', 'Variable', 'Unit Code', 'Unit',
       'PowerCode Code', 'PowerCode', 'Reference Period Code',
       'Reference Period', 'Value', 'Flag Codes', 'Flags'],
      dtype='object')

In [221]:
# We are gonna merge population and pm25

df1=population_new.merge(pm25, left_on='Country', right_on='Country')
df1

Rank                    Country  Population Date_of_Information  COU  \
0       2                      India  1399179585           2023 est.  IND   
1       3              United States   339665118           2023 est.  USA   
2       4                  Indonesia   279476346           2023 est.  IDN   
3       5                   Pakistan   247653551           2023 est.  PAK   
4       6                    Nigeria   230842743           2023 est.  NGA   
..    ...                        ...         ...                 ...  ...   
195   227  Saint Pierre and Miquelon        5195           2023 est.  SPM   
196   231                       Niue        2000      July 2022 est.  NIU   
197   232             Norfolk Island        1748           2016 est.  NFK   
198   233                    Tokelau        1647           2019 est.  TKL   
199   236    Cocos (Keeling) Islands         596      July 2014 est.  CCK   

     SMALL_SUBNATIONAL_REGION Small subnational region  \
0                         NaN           Not applicable   
1                         NaN           Not applicable   
2                         NaN           Not applicable   
3                         NaN           Not applicable   
4                         NaN           Not applicable   
..                        ...                      ...   
195                       NaN           Not applicable   
196                       NaN           Not applicable   
197                       NaN           Not applicable   
198                       NaN           Not applicable   
199                       NaN           Not applicable   

    LARGE_SUBNATIONAL_REGION Large subnational region     VAR  \
0                      TOTAL                    Total  PWM_EX   
1                      TOTAL                    Total  PWM_EX   
2                      TOTAL                    Total  PWM_EX   
3                      TOTAL                    Total  PWM_EX   
4                      TOTAL                    Total  PWM_EX   
..                       ...                      ...     ...   
195                    TOTAL                    Total  PWM_EX   
196                    TOTAL                    Total  PWM_EX   
197                    TOTAL                    Total  PWM_EX   
198                    TOTAL                    Total  PWM_EX   
199                    TOTAL                    Total  PWM_EX   

                              Variable Unit Code                        Unit  \
0    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
1    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
2    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
3    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
4    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
..                                 ...       ...                         ...   
195  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
196  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
197  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
198  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
199  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   

     PowerCode Code PowerCode  Reference Period Code  Reference Period  \
0                 0     Units                    NaN               NaN   
1                 0     Units                    NaN               NaN   
2                 0     Units                    NaN               NaN   
3                 0     Units                    NaN               NaN   
4                 0     Units                    NaN               NaN   
..              ...       ...                    ...               ...   
195               0     Units                    NaN               NaN   
196               0     Units                    

In [222]:
# We are gonna merge the previous merge and h2O
df2=df1.merge(h2O, left_on='Country', right_on='Entity')
df2

Rank                           Country  Population Date_of_Information  \
0       2                             India  1399179585           2023 est.   
1       3                     United States   339665118           2023 est.   
2       4                         Indonesia   279476346           2023 est.   
3       5                          Pakistan   247653551           2023 est.   
4       6                           Nigeria   230842743           2023 est.   
..    ...                               ...         ...                 ...   
157   195  Saint Vincent and the Grenadines      100804           2023 est.   
158   197                        Seychelles       97617           2023 est.   
159   201                          Dominica       74656           2023 est.   
160   207             Saint Kitts and Nevis       54817           2023 est.   
161   216                            Monaco       31597           2023 est.   

     COU  SMALL_SUBNATIONAL_REGION Small subnational region  \
0    IND                       NaN           Not applicable   
1    USA                       NaN           Not applicable   
2    IDN                       NaN           Not applicable   
3    PAK                       NaN           Not applicable   
4    NGA                       NaN           Not applicable   
..   ...                       ...                      ...   
157  VCT                       NaN           Not applicable   
158  SYC                       NaN           Not applicable   
159  DMA                       NaN           Not applicable   
160  KNA                       NaN           Not applicable   
161  MCO                       NaN           Not applicable   

    LARGE_SUBNATIONAL_REGION Large subnational region     VAR  ...  \
0                      TOTAL                    Total  PWM_EX  ...   
1                      TOTAL                    Total  PWM_EX  ...   
2                      TOTAL                    Total  PWM_EX  ...   
3                      TOTAL                    Total  PWM_EX  ...   
4                      TOTAL                    Total  PWM_EX  ...   
..                       ...                      ...     ...  ...   
157                    TOTAL                    Total  PWM_EX  ...   
158                    TOTAL                    Total  PWM_EX  ...   
159                    TOTAL                    Total  PWM_EX  ...   
160                    TOTAL                    Total  PWM_EX  ...   
161                    TOTAL                    Total  PWM_EX  ...   

    PowerCode Code PowerCode Reference Period Code  Reference Period  Value  \
0                0     Units                   NaN               NaN  68.12   
1                0     Units                   NaN               NaN   7.84   
2                0     Units                   NaN               NaN  18.69   
3                0     Units                   NaN               NaN  64.30   
4                0     Units                   NaN               NaN  57.04   
..             ...       ...                   ...               ...    ...   
157              0     Units                   NaN               NaN  23.62   
158              0     Units                   NaN               NaN   8.98   
159              0     Units                   NaN               NaN  20.15   
160              0     Units                   NaN               NaN   7.54   
161              0     Units                   NaN               NaN  10.70   

     Flag Codes            Flags                            Entity Code  \
0             E  Estimated value                             India  IND   
1             E  Estimated value                     United States  USA   
2             E  Estimated value                         Indonesia  IDN   
3             E  Estimated value                          Pakistan  PAK   
4             E  Estimated value                           Nigeria  NGA   
..          ...              ...                           

In [227]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162 entries, 0 to 161
Data columns (total 23 columns):
 #   Column                                                       Non-Null Count  Dtype  
---  ------                                                       --------------  -----  
 0   Rank                                                         162 non-null    int64  
 1   Country                                                      162 non-null    object 
 2   Population                                                   162 non-null    int64  
 3   Date_of_Information                                          162 non-null    object 
 4   COU                                                          162 non-null    object 
 5   SMALL_SUBNATIONAL_REGION                                     0 non-null      float64
 6   Small subnational region                                     162 non-null    object 
 7   LARGE_SUBNATIONAL_REGION                                     162 non-null    obj

We have lost several countries. Then we will try fuzzy merge

In [232]:
#pm25.columns

Index(['COU', 'Country', 'SMALL_SUBNATIONAL_REGION',
       'Small subnational region', 'LARGE_SUBNATIONAL_REGION',
       'Large subnational region', 'VAR', 'Variable', 'Unit Code', 'Unit',
       'PowerCode Code', 'PowerCode', 'Reference Period Code',
       'Reference Period', 'Value', 'Flag Codes', 'Flags'],
      dtype='object')

In [233]:
OnlyPopulation=set(population_new.Country)-set(pm25.Country)
OnlyPm25=set(pm25.Country)-set(population_new.Country)

from thefuzz import process as fz
[(f,fz.extractOne(f, OnlyPopulation)) for f in sorted(OnlyPm25)]

[('Bahamas', ('Bahamas, The', 90)),
 ('Brunei Darussalam', ('Brunei', 90)),
 ("China (People's Republic of)", ('China', 90)),
 ('Chinese Taipei', ('China', 72)),
 ('Christmas Islands', ('Christmas Island', 97)),
 ('Congo', ('Congo, Republic of the', 90)),
 ('Czech Republic', ('Congo, Republic of the', 86)),
 ("Côte d'Ivoire", ("Cote d'Ivoire", 96)),
 ("Democratic People's Republic of Korea", ('Korea, South', 86)),
 ('Democratic Republic of the Congo', ('Congo, Republic of the', 95)),
 ('Euro area (19 countries)', ('Korea, South', 55)),
 ('European Union (28 countries)', ('Korea, North', 47)),
 ('European Union – 27 countries (from 01/02/2020)', ('Korea, North', 45)),
 ('Faeroe Islands', ('Faroe Islands', 96)),
 ('Falkland Islands (Malvinas)', ('Falkland Islands (Islas Malvinas)', 95)),
 ('G20', ('Congo, Republic of the', 30)),
 ('G7', ('Gaza Strip', 45)),
 ('Gambia', ('Gambia, The', 90)),
 ('Holy See', ('Holy See (Vatican City)', 90)),
 ('Hong Kong, China', ('Hong Kong', 90)),
 ('Korea

In [234]:
[(f,fz.extractOne(f, OnlyPopulation)) for f in sorted(OnlyPm25)
 if fz.extractOne(f, OnlyPopulation)[1]>=87]

[('Bahamas', ('Bahamas, The', 90)),
 ('Brunei Darussalam', ('Brunei', 90)),
 ("China (People's Republic of)", ('China', 90)),
 ('Christmas Islands', ('Christmas Island', 97)),
 ('Congo', ('Congo, Republic of the', 90)),
 ("Côte d'Ivoire", ("Cote d'Ivoire", 96)),
 ('Democratic Republic of the Congo', ('Congo, Republic of the', 95)),
 ('Faeroe Islands', ('Faroe Islands', 96)),
 ('Falkland Islands (Malvinas)', ('Falkland Islands (Islas Malvinas)', 95)),
 ('Gambia', ('Gambia, The', 90)),
 ('Holy See', ('Holy See (Vatican City)', 90)),
 ('Hong Kong, China', ('Hong Kong', 90)),
 ('Korea', ('Korea, South', 90)),
 ('Macau, China', ('China', 90)),
 ('Micronesia', ('Micronesia, Federated States of', 90)),
 ('Palestinian Authority or West Bank and Gaza Strip', ('West Bank', 90)),
 ('Pitcairn', ('Pitcairn Islands', 90)),
 ('Saint Helena', ('Saint Helena, Ascension, and Tristan da Cunha', 90)),
 ('South Sudan ', ('South Sudan', 100)),
 ('Syrian Arab Republic', ('Syria', 90)),
 ('United States Virgi

In [239]:
# We have to create a dictionary
changesPm25_1={f:fz.extractOne(f, OnlyPopulation)[0] 
                 for f in sorted(OnlyPm25)
                 if fz.extractOne(f, OnlyPopulation)[1] >=87}
#dict of matches
changesPm25_1

{'Bahamas': 'Bahamas, The',
 'Brunei Darussalam': 'Brunei',
 "China (People's Republic of)": 'China',
 'Christmas Islands': 'Christmas Island',
 'Congo': 'Congo, Republic of the',
 "Côte d'Ivoire": "Cote d'Ivoire",
 'Democratic Republic of the Congo': 'Congo, Republic of the',
 'Faeroe Islands': 'Faroe Islands',
 'Falkland Islands (Malvinas)': 'Falkland Islands (Islas Malvinas)',
 'Gambia': 'Gambia, The',
 'Holy See': 'Holy See (Vatican City)',
 'Hong Kong, China': 'Hong Kong',
 'Korea': 'Korea, South',
 'Macau, China': 'China',
 'Micronesia': 'Micronesia, Federated States of',
 'Palestinian Authority or West Bank and Gaza Strip': 'West Bank',
 'Pitcairn': 'Pitcairn Islands',
 'Saint Helena': 'Saint Helena, Ascension, and Tristan da Cunha',
 'South Sudan ': 'South Sudan',
 'Syrian Arab Republic': 'Syria',
 'United States Virgin Islands': 'Virgin Islands',
 'Viet Nam': 'Vietnam'}

In [241]:
pm25.Country.replace(to_replace=changesPm25_1,inplace=True)

COU                                          Country  \
0          AUS                                        Australia   
2          AUT                                          Austria   
4          BEL                                          Belgium   
6          CAN                                           Canada   
8          CZE                                   Czech Republic   
..         ...                                              ...   
468       OECD                                     OECD - Total   
470      OECDE                                    OECD - Europe   
472       EA19                         Euro area (19 countries)   
474  EU27_2020  European Union – 27 countries (from 01/02/2020)   
476        MYT                                          Mayotte   

     SMALL_SUBNATIONAL_REGION Small subnational region  \
0                         NaN           Not applicable   
2                         NaN           Not applicable   
4                         NaN           Not applicable   
6                         NaN           Not applicable   
8                         NaN           Not applicable   
..                        ...                      ...   
468                       NaN           Not applicable   
470                       NaN           Not applicable   
472                       NaN           Not applicable   
474                       NaN           Not applicable   
476                       NaN           Not applicable   

    LARGE_SUBNATIONAL_REGION Large subnational region     VAR  \
0                      TOTAL                    Total  PWM_EX   
2                      TOTAL                    Total  PWM_EX   
4                      TOTAL                    Total  PWM_EX   
6                      TOTAL                    Total  PWM_EX   
8                      TOTAL                    Total  PWM_EX   
..                       ...                      ...     ...   
468                    TOTAL                    Total  PWM_EX   
470                    TOTAL                    Total  PWM_EX   
472                    TOTAL                    Total  PWM_EX   
474                    TOTAL                    Total  PWM_EX   
476                    TOTAL                    Total  PWM_EX   

                              Variable Unit Code                        Unit  \
0    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
2    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
4    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
6    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
8    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
..                                 ...       ...                         ...   
468  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
470  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
472  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
474  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
476  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   

     PowerCode Code PowerCode  Reference Period Code  Reference Period  \
0                 0     Units                    NaN               NaN   
2                 0     Units                    NaN               NaN   
4                 0     Units                    NaN               NaN   
6                 0     Units                    NaN               NaN   
8                 0     Units                    NaN               NaN   
..              ...       ...                    ...               ...   
468               0     Units                    NaN               NaN   
470               0     Units                    NaN               NaN   
472               0     Units                    NaN               NaN   
474               0     Uni

In [242]:
# second try
OnlyPopulation=set(population_new.Country)-set(pm25.Country)
OnlyPm25=set(pm25.Country)-set(population_new.Country)
[(f,fz.extractOne(f, OnlyPopulation)) for f in sorted(OnlyPm25)]

[('Chinese Taipei', ('Taiwan', 45)),
 ('Czech Republic', ('Congo, Democratic Republic of the', 86)),
 ("Democratic People's Republic of Korea", ('Korea, North', 86)),
 ('Euro area (19 countries)', ('Burma', 54)),
 ('European Union (28 countries)', ('Korea, North', 47)),
 ('European Union – 27 countries (from 01/02/2020)', ('Korea, North', 45)),
 ('G20', ('Gaza Strip', 30)),
 ('G7', ('Gaza Strip', 45)),
 ("Lao People's Democratic Republic",
  ('Congo, Democratic Republic of the', 71)),
 ('Mayotte', ('Sint Maarten', 51)),
 ('Myanmar', ('Saint Martin', 53)),
 ('Netherlands Antilles', ('Saint Martin', 45)),
 ('OECD - Europe', ('Turkey (Turkiye)', 43)),
 ('OECD - Total', ('Korea, North', 50)),
 ('Slovak Republic', ('Congo, Democratic Republic of the', 86)),
 ('Türkiye', ('Turkey (Turkiye)', 75)),
 ('World', ('Congo, Democratic Republic of the', 36))]

In [244]:
# second dict of changes 
# we select a different threshold
changesPm25_2={f:fz.extractOne(f, OnlyPopulation)[0] 
                 for f in sorted(OnlyPm25)
                 if 74<=fz.extractOne(f, OnlyPopulation)[1]}

#dict of matches
changesPm25_2

{'Czech Republic': 'Congo, Democratic Republic of the',
 "Democratic People's Republic of Korea": 'Korea, North',
 'Slovak Republic': 'Congo, Democratic Republic of the',
 'Türkiye': 'Turkey (Turkiye)'}

In [245]:
pm25.Country.replace(to_replace=changesPm25_2,inplace=True)

COU                                          Country  \
0          AUS                                        Australia   
2          AUT                                          Austria   
4          BEL                                          Belgium   
6          CAN                                           Canada   
8          CZE                Congo, Democratic Republic of the   
..         ...                                              ...   
468       OECD                                     OECD - Total   
470      OECDE                                    OECD - Europe   
472       EA19                         Euro area (19 countries)   
474  EU27_2020  European Union – 27 countries (from 01/02/2020)   
476        MYT                                          Mayotte   

     SMALL_SUBNATIONAL_REGION Small subnational region  \
0                         NaN           Not applicable   
2                         NaN           Not applicable   
4                         NaN           Not applicable   
6                         NaN           Not applicable   
8                         NaN           Not applicable   
..                        ...                      ...   
468                       NaN           Not applicable   
470                       NaN           Not applicable   
472                       NaN           Not applicable   
474                       NaN           Not applicable   
476                       NaN           Not applicable   

    LARGE_SUBNATIONAL_REGION Large subnational region     VAR  \
0                      TOTAL                    Total  PWM_EX   
2                      TOTAL                    Total  PWM_EX   
4                      TOTAL                    Total  PWM_EX   
6                      TOTAL                    Total  PWM_EX   
8                      TOTAL                    Total  PWM_EX   
..                       ...                      ...     ...   
468                    TOTAL                    Total  PWM_EX   
470                    TOTAL                    Total  PWM_EX   
472                    TOTAL                    Total  PWM_EX   
474                    TOTAL                    Total  PWM_EX   
476                    TOTAL                    Total  PWM_EX   

                              Variable Unit Code                        Unit  \
0    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
2    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
4    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
6    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
8    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
..                                 ...       ...                         ...   
468  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
470  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
472  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
474  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
476  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   

     PowerCode Code PowerCode  Reference Period Code  Reference Period  \
0                 0     Units                    NaN               NaN   
2                 0     Units                    NaN               NaN   
4                 0     Units                    NaN               NaN   
6                 0     Units                    NaN               NaN   
8                 0     Units                    NaN               NaN   
..              ...       ...                    ...               ...   
468               0     Units                    NaN               NaN   
470               0     Units                    NaN               NaN   
472               0     Units                    NaN               NaN   
474               0     Uni

In [246]:
# third try
OnlyPopulation=set(population_new.Country)-set(pm25.Country)
OnlyPm25=set(pm25.Country)-set(population_new.Country)
[(f,fz.extractOne(f, OnlyPopulation)) for f in sorted(OnlyPm25)]

[('Chinese Taipei', ('Taiwan', 45)),
 ('Euro area (19 countries)', ('Burma', 54)),
 ('European Union (28 countries)', ('Sint Maarten', 43)),
 ('European Union – 27 countries (from 01/02/2020)', ('Curacao', 39)),
 ('G20', ('Gaza Strip', 30)),
 ('G7', ('Gaza Strip', 45)),
 ("Lao People's Democratic Republic", ('Laos', 68)),
 ('Mayotte', ('Sint Maarten', 51)),
 ('Myanmar', ('Saint Martin', 53)),
 ('Netherlands Antilles', ('Saint Martin', 45)),
 ('OECD - Europe', ('Curacao', 41)),
 ('OECD - Total', ('Czechia', 32)),
 ('World', ('Paracel Islands', 36))]

In [247]:
# third dich of changes
# new threshold
changesPm25_3={f:fz.extractOne(f, OnlyPopulation)[0] 
                 for f in sorted(OnlyPm25)
                 if 54==fz.extractOne(f, OnlyPopulation)[1]}
# dict of matches
changesPm25_3

{'Euro area (19 countries)': 'Burma'}

In [251]:
pm25.Country.replace(to_replace=changesPm25_3,inplace=True)
pm25

COU                                          Country  \
0          AUS                                        Australia   
2          AUT                                          Austria   
4          BEL                                          Belgium   
6          CAN                                           Canada   
8          CZE                Congo, Democratic Republic of the   
..         ...                                              ...   
468       OECD                                     OECD - Total   
470      OECDE                                    OECD - Europe   
472       EA19                                            Burma   
474  EU27_2020  European Union – 27 countries (from 01/02/2020)   
476        MYT                                          Mayotte   

     SMALL_SUBNATIONAL_REGION Small subnational region  \
0                         NaN           Not applicable   
2                         NaN           Not applicable   
4                         NaN           Not applicable   
6                         NaN           Not applicable   
8                         NaN           Not applicable   
..                        ...                      ...   
468                       NaN           Not applicable   
470                       NaN           Not applicable   
472                       NaN           Not applicable   
474                       NaN           Not applicable   
476                       NaN           Not applicable   

    LARGE_SUBNATIONAL_REGION Large subnational region     VAR  \
0                      TOTAL                    Total  PWM_EX   
2                      TOTAL                    Total  PWM_EX   
4                      TOTAL                    Total  PWM_EX   
6                      TOTAL                    Total  PWM_EX   
8                      TOTAL                    Total  PWM_EX   
..                       ...                      ...     ...   
468                    TOTAL                    Total  PWM_EX   
470                    TOTAL                    Total  PWM_EX   
472                    TOTAL                    Total  PWM_EX   
474                    TOTAL                    Total  PWM_EX   
476                    TOTAL                    Total  PWM_EX   

                              Variable Unit Code                        Unit  \
0    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
2    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
4    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
6    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
8    Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
..                                 ...       ...                         ...   
468  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
470  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
472  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
474  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   
476  Mean population exposure to PM2.5  MICRO_M3  Micrograms per cubic metre   

     PowerCode Code PowerCode  Reference Period Code  Reference Period  \
0                 0     Units                    NaN               NaN   
2                 0     Units                    NaN               NaN   
4                 0     Units                    NaN               NaN   
6                 0     Units                    NaN               NaN   
8                 0     Units                    NaN               NaN   
..              ...       ...                    ...               ...   
468               0     Units                    NaN               NaN   
470               0     Units                    NaN               NaN   
472               0     Units                    NaN               NaN   
474               0     Uni

In [252]:
OnlyPopulation=set(population_new.Country)-set(pm25.Country)
OnlyPm25=set(pm25.Country)-set(population_new.Country)
[(f,fz.extractOne(f, OnlyPopulation)) for f in sorted(OnlyPm25)]

[('Chinese Taipei', ('Taiwan', 45)),
 ('European Union (28 countries)', ('Sint Maarten', 43)),
 ('European Union – 27 countries (from 01/02/2020)', ('Curacao', 39)),
 ('G20', ('Gaza Strip', 30)),
 ('G7', ('Gaza Strip', 45)),
 ("Lao People's Democratic Republic", ('Laos', 68)),
 ('Mayotte', ('Sint Maarten', 51)),
 ('Myanmar', ('Saint Martin', 53)),
 ('Netherlands Antilles', ('Saint Martin', 45)),
 ('OECD - Europe', ('Curacao', 41)),
 ('OECD - Total', ('Czechia', 32)),
 ('World', ('Paracel Islands', 36))]

The fourth attempt did not offer good results. So we are ready

In [253]:
df1=pm25.merge(population_new)
df1

COU                            Country  SMALL_SUBNATIONAL_REGION  \
0     AUS                          Australia                       NaN   
1     AUT                            Austria                       NaN   
2     BEL                            Belgium                       NaN   
3     CAN                             Canada                       NaN   
4     CZE  Congo, Democratic Republic of the                       NaN   
..    ...                                ...                       ...   
222   GGY                           Guernsey                       NaN   
223   IMN                        Isle of Man                       NaN   
224   JEY                             Jersey                       NaN   
225   SSD                        South Sudan                       NaN   
226  EA19                              Burma                       NaN   

    Small subnational region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
222           Not applicable                    TOTAL   
223           Not applicable                    TOTAL   
224           Not applicable                    TOTAL   
225           Not applicable                    TOTAL   
226           Not applicable                    TOTAL   

    Large subnational region     VAR                           Variable  \
0                      Total  PWM_EX  Mean population exposure to PM2.5   
1                      Total  PWM_EX  Mean population exposure to PM2.5   
2                      Total  PWM_EX  Mean population exposure to PM2.5   
3                      Total  PWM_EX  Mean population exposure to PM2.5   
4                      Total  PWM_EX  Mean population exposure to PM2.5   
..                       ...     ...                                ...   
222                    Total  PWM_EX  Mean population exposure to PM2.5   
223                    Total  PWM_EX  Mean population exposure to PM2.5   
224                    Total  PWM_EX  Mean population exposure to PM2.5   
225                    Total  PWM_EX  Mean population exposure to PM2.5   
226                    Total  PWM_EX  Mean population exposure to PM2.5   

    Unit Code                        Unit  PowerCode Code PowerCode  \
0    MICRO_M3  Micrograms per cubic metre               0     Units   
1    MICRO_M3  Micrograms per cubic metre               0     Units   
2    MICRO_M3  Micrograms per cubic metre               0     Units   
3    MICRO_M3  Micrograms per cubic metre               0     Units   
4    MICRO_M3  Micrograms per cubic metre               0     Units   
..        ...                         ...             ...       ...   
222  MICRO_M3  Micrograms per cubic metre               0     Units   
223  MICRO_M3  Micrograms per cubic metre               0     Units   
224  MICRO_M3  Micrograms per cubic metre               0     Units   
225  MICRO_M3  Micrograms per cubic metre               0     Units   
226  MICRO_M3  Micrograms per cubic metre               0     Units   

     Reference Period Code  Reference Period   Value Flag Codes  \
0                      NaN               NaN   7.730          E   
1                      NaN               NaN  12.720          E   
2                      NaN               NaN  12.540          E   
3                      NaN               NaN   7.090          E   
4                      NaN               NaN  17.560          E   
..                     ...               ...     ...        ...   
222                    NaN               NaN  10.980          E   
223                    NaN               NaN   9.030          E   
224                    NaN               NaN  11.130          E   
225     

In [255]:
# Now we do fuzzy merge for the merged df "df1" and the data frame of h2O
OnlyDf1=set(df1.Country)-set(h2O.Entity)
OnlyH2O=set(h2O.Entity)-set(df1.Country)

[(f,fz.extractOne(f, OnlyH2O)) for f in sorted(OnlyDf1)]

[('American Samoa', ('Latin America and Caribbean (WB)', 64)),
 ('Andorra', ('Middle East and North Africa (WB)', 64)),
 ('Anguilla', ('Laos', 45)),
 ('Aruba', ('Gambia', 55)),
 ('Bahamas, The', ('Gambia', 45)),
 ('Bermuda', ('Lower-middle-income countries', 51)),
 ('British Virgin Islands', ('Laos', 45)),
 ('Burma', ('South Korea', 38)),
 ('Cabo Verde', ('Cape Verde', 80)),
 ('Cayman Islands', ('Gambia', 45)),
 ('Christmas Island', ('Laos', 45)),
 ('Cocos (Keeling) Islands', ('Congo', 54)),
 ('Congo, Democratic Republic of the', ('Democratic Republic of Congo', 95)),
 ('Congo, Republic of the', ('Congo', 90)),
 ('Cook Islands', ('Congo', 54)),
 ('Falkland Islands (Islas Malvinas)', ('Gambia', 45)),
 ('Faroe Islands', ('Europe and Central Asia (WB)', 49)),
 ('French Polynesia', ('North Korea', 44)),
 ('Gambia, The', ('Gambia', 90)),
 ('Gibraltar', ('Latin America and Caribbean (WB)', 48)),
 ('Greenland', ('Europe and Central Asia (WB)', 50)),
 ('Guam', ('Gambia', 68)),
 ('Guernsey', ('

We prepare the changes: 

In [256]:
# first change
changesDf1_1={f:fz.extractOne(f, OnlyH2O)[0] 
                 for f in sorted(OnlyDf1)
                 if fz.extractOne(f, OnlyH2O)[1] >=87}
# dict of matches
changesDf1_1

{'Congo, Democratic Republic of the': 'Democratic Republic of Congo',
 'Congo, Republic of the': 'Congo',
 'Gambia, The': 'Gambia',
 'Korea, North': 'North Korea',
 'Korea, South': 'South Korea',
 'Turkey (Turkiye)': 'Turkey',
 'West Bank': 'West Bank and Gaza'}

In [260]:
# make changes
df1.Country.replace(to_replace=changesDf1_1,inplace=True)

We do the second iteration

In [258]:
OnlyDf1=set(df1.Country)-set(h2O.Entity)
OnlyH2O=set(h2O.Entity)-set(df1.Country)

[(f,fz.extractOne(f, OnlyH2O)) for f in sorted(OnlyDf1)]

[('American Samoa', ('Latin America and Caribbean (WB)', 64)),
 ('Andorra', ('Middle East and North Africa (WB)', 64)),
 ('Anguilla', ('Laos', 45)),
 ('Aruba', ('Sub-Saharan Africa (WB)', 54)),
 ('Bahamas, The', ('Laos', 45)),
 ('Bermuda', ('Lower-middle-income countries', 51)),
 ('British Virgin Islands', ('Laos', 45)),
 ('Burma', ('North America (WB)', 36)),
 ('Cabo Verde', ('Cape Verde', 80)),
 ('Cayman Islands', ('Laos', 45)),
 ('Christmas Island', ('Laos', 45)),
 ('Cocos (Keeling) Islands', ('Laos', 45)),
 ('Cook Islands', ('Laos', 45)),
 ('Falkland Islands (Islas Malvinas)', ('Laos', 45)),
 ('Faroe Islands', ('Europe and Central Asia (WB)', 49)),
 ('French Polynesia', ('Czechia', 43)),
 ('Gibraltar', ('Latin America and Caribbean (WB)', 48)),
 ('Greenland', ('Europe and Central Asia (WB)', 50)),
 ('Guam', ('North America (WB)', 45)),
 ('Guernsey', ('High-income countries', 34)),
 ('Holy See (Vatican City)', ('Laos', 45)),
 ('Hong Kong', ('High-income countries', 40)),
 ('Isle of 

In [261]:
# second dict of changes 
# we select a different threshold
changesDf1_2={f:fz.extractOne(f, OnlyH2O)[0] 
                 for f in sorted(OnlyDf1)
                 if 74<=fz.extractOne(f, OnlyH2O)[1]}

#dict of matches
changesPm25_2

{'Czech Republic': 'Congo, Democratic Republic of the',
 "Democratic People's Republic of Korea": 'Korea, North',
 'Slovak Republic': 'Congo, Democratic Republic of the',
 'Türkiye': 'Turkey (Turkiye)'}

In [262]:
df1.Country.replace(to_replace=changesPm25_2,inplace=True)

In [263]:
# third change
OnlyDf1=set(df1.Country)-set(h2O.Entity)
OnlyH2O=set(h2O.Entity)-set(df1.Country)

[(f,fz.extractOne(f, OnlyH2O)) for f in sorted(OnlyDf1)]

[('American Samoa', ('Latin America and Caribbean (WB)', 64)),
 ('Andorra', ('Middle East and North Africa (WB)', 64)),
 ('Anguilla', ('Laos', 45)),
 ('Aruba', ('Sub-Saharan Africa (WB)', 54)),
 ('Bahamas, The', ('Laos', 45)),
 ('Bermuda', ('Lower-middle-income countries', 51)),
 ('British Virgin Islands', ('Laos', 45)),
 ('Burma', ('North America (WB)', 36)),
 ('Cabo Verde', ('Cape Verde', 80)),
 ('Cayman Islands', ('Laos', 45)),
 ('Christmas Island', ('Laos', 45)),
 ('Cocos (Keeling) Islands', ('Laos', 45)),
 ('Cook Islands', ('Laos', 45)),
 ('Falkland Islands (Islas Malvinas)', ('Laos', 45)),
 ('Faroe Islands', ('Europe and Central Asia (WB)', 49)),
 ('French Polynesia', ('Czechia', 43)),
 ('Gibraltar', ('Latin America and Caribbean (WB)', 48)),
 ('Greenland', ('Europe and Central Asia (WB)', 50)),
 ('Guam', ('North America (WB)', 45)),
 ('Guernsey', ('High-income countries', 34)),
 ('Holy See (Vatican City)', ('Laos', 45)),
 ('Hong Kong', ('High-income countries', 40)),
 ('Isle of 

In [265]:
# third dict of changes 
# we select a different threshold
changesDf1_3={f:fz.extractOne(f, OnlyH2O)[0] 
                 for f in sorted(OnlyDf1)
                 if 54==fz.extractOne(f, OnlyH2O)[1]}

#dict of matches
changesPm25_3

{'Euro area (19 countries)': 'Burma'}

In [266]:
df2=df1.merge(h2O, left_on='Country', right_on= 'Entity', how='inner')
df2

COU                       Country  SMALL_SUBNATIONAL_REGION  \
0    AUS                     Australia                       NaN   
1    AUT                       Austria                       NaN   
2    BEL                       Belgium                       NaN   
3    CAN                        Canada                       NaN   
4    CZE  Democratic Republic of Congo                       NaN   
..   ...                           ...                       ...   
173  ZMB                        Zambia                       NaN   
174  ZWE                      Zimbabwe                       NaN   
175  SRB                        Serbia                       NaN   
176  MNE                    Montenegro                       NaN   
177  SSD                   South Sudan                       NaN   

    Small subnational region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
173           Not applicable                    TOTAL   
174           Not applicable                    TOTAL   
175           Not applicable                    TOTAL   
176           Not applicable                    TOTAL   
177           Not applicable                    TOTAL   

    Large subnational region     VAR                           Variable  \
0                      Total  PWM_EX  Mean population exposure to PM2.5   
1                      Total  PWM_EX  Mean population exposure to PM2.5   
2                      Total  PWM_EX  Mean population exposure to PM2.5   
3                      Total  PWM_EX  Mean population exposure to PM2.5   
4                      Total  PWM_EX  Mean population exposure to PM2.5   
..                       ...     ...                                ...   
173                    Total  PWM_EX  Mean population exposure to PM2.5   
174                    Total  PWM_EX  Mean population exposure to PM2.5   
175                    Total  PWM_EX  Mean population exposure to PM2.5   
176                    Total  PWM_EX  Mean population exposure to PM2.5   
177                    Total  PWM_EX  Mean population exposure to PM2.5   

    Unit Code                        Unit  ...  Reference Period   Value  \
0    MICRO_M3  Micrograms per cubic metre  ...               NaN   7.730   
1    MICRO_M3  Micrograms per cubic metre  ...               NaN  12.720   
2    MICRO_M3  Micrograms per cubic metre  ...               NaN  12.540   
3    MICRO_M3  Micrograms per cubic metre  ...               NaN   7.090   
4    MICRO_M3  Micrograms per cubic metre  ...               NaN  17.560   
..        ...                         ...  ...               ...     ...   
173  MICRO_M3  Micrograms per cubic metre  ...               NaN  23.360   
174  MICRO_M3  Micrograms per cubic metre  ...               NaN  19.160   
175  MICRO_M3  Micrograms per cubic metre  ...               NaN  26.073   
176  MICRO_M3  Micrograms per cubic metre  ...               NaN  23.260   
177  MICRO_M3  Micrograms per cubic metre  ...               NaN  38.849   

     Flag Codes            Flags  Rank Population Date_of_Information  \
0             E  Estimated value    54   26461166           2023 est.   
1             E  Estimated value    99    8940860           2023 est.   
2             E  Estimated value    82   11913633           2023 est.   
3             E  Estimated value    38   38516736           2023 est.   
4             E  Estimated value    14  111859928           2023 est.   
..          ...              ...   ...        ...                 ...   
173           E  Estimated value    63   20216029           2023 est.   
174           E  Estimated value    73   15418674           2023 est.   
175         NaN              NaN

In [267]:
import os
country_isos=pd.read_excel(os.path.join("data","country_isos.xlsx"))
country_isos

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
166           UZBEKISTAN                   The Republic of Uzbekistan   
167             VIET NAM           The Socialist Republic of Viet Nam   
168                YEMEN                        The Republic of Yemen   
169               ZAMBIA                       The Republic of Zambia   
170             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  
0             .af   AF  AFG  
1             .al   AL  ALB  
2             .dz   DZ  DZA  
3             .ao   AO  AGO  
4             .ag   AG  ATG  
..            ...  ...  ...  
166           .uz   UZ  UZB  
167           .vn   VN  VNM  
168           .ye   YE  YEM  
169           .zm   ZM  ZMB  
170           .zw   ZW  ZWE  

[171 rows x 5 columns]

In [269]:
df2.Country=df2.Country.str.upper()
country_isos.merge(df2, left_on='Country',right_on='Country')

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
135              URUGUAY             The Oriental Republic of Uruguay   
136           UZBEKISTAN                   The Republic of Uzbekistan   
137                YEMEN                        The Republic of Yemen   
138               ZAMBIA                       The Republic of Zambia   
139             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  COU  SMALL_SUBNATIONAL_REGION  \
0             .af   AF  AFG  AFG                       NaN   
1             .al   AL  ALB  ALB                       NaN   
2             .dz   DZ  DZA  DZA                       NaN   
3             .ao   AO  AGO  AGO                       NaN   
4             .ag   AG  ATG  ATG                       NaN   
..            ...  ...  ...  ...                       ...   
135           .uy   UY  URY  URY                       NaN   
136           .uz   UZ  UZB  UZB                       NaN   
137           .ye   YE  YEM  YEM                       NaN   
138           .zm   ZM  ZMB  ZMB                       NaN   
139           .zw   ZW  ZWE  ZWE                       NaN   

    Small subnational region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
135           Not applicable                    TOTAL   
136           Not applicable                    TOTAL   
137           Not applicable                    TOTAL   
138           Not applicable                    TOTAL   
139           Not applicable                    TOTAL   

    Large subnational region  ... Reference Period  Value Flag Codes  \
0                      Total  ...              NaN  68.27          E   
1                      Total  ...              NaN  18.90          E   
2                      Total  ...              NaN  24.93          E   
3                      Total  ...              NaN  24.20          E   
4                      Total  ...              NaN  18.99          E   
..                       ...  ...              ...    ...        ...   
135                    Total  ...              NaN  10.16          E   
136                    Total  ...              NaN  36.29          E   
137                    Total  ...              NaN  44.03          E   
138                    Total  ...              NaN  23.36          E   
139                    Total  ...              NaN  19.16          E   

               Flags  Rank Population  Date_of_Information  \
0    Estimated value    37   39232003            2023 est.   
1    Estimated value   137    3101621            2023 est.   
2    Estimated value    34   44758398            2023 est.   
3    Estimated value    41   35981281            2023 est.   
4    Estimated value   194     101489            2023 est.   
..               ...   ...        ...                  ...   
135  Estimated value   132    3416264            2023 est.   
136  Estimated value    48   31360836            2023 est.   
137  Estimated value    47   31565602            2023 est.   
138  Estimated value    63   20216029            2023 est.   
139  Estimated value    73   15418674            2023 est.   

                  Entity  Code  \
0            Afghanistan   AFG   
1                Albania   ALB  

In [271]:
# We have lost several countries, then we do the fuzzy merge
OnlyDf2=set(df2.Country)-set(country_isos.Country)
OnlyISOS=set(country_isos.Country)-set(df2.Country)

In [272]:
from thefuzz import process as fz
[(f,fz.extractOne(f, OnlyISOS)) for f in sorted(OnlyDf2)]

[('BOLIVIA', ('BOLIVIA (PLURINATIONAL STATE OF)', 90)),
 ('BRUNEI', ('BRUNEI DARUSSALAM', 90)),
 ('CENTRAL AFRICAN REPUBLIC', ('CENTRAL AFRICAN REPUBLIC (THE)', 95)),
 ('COMOROS', ('COMOROS (THE)', 90)),
 ('CONGO', ('CONGO (THE)', 90)),
 ("COTE D'IVOIRE", ("CÔTE D'IVOIRE", 92)),
 ('DEMOCRATIC REPUBLIC OF CONGO',
  ('CONGO (THE DEMOCRATIC REPUBLIC OF THE)', 95)),
 ('DOMINICA', ('DOMINICAN REPUBLIC (THE)', 90)),
 ('DOMINICAN REPUBLIC', ('DOMINICAN REPUBLIC (THE)', 95)),
 ('ERITREA', ('UNITED ARAB EMIRATES (THE)', 51)),
 ('GAMBIA', ('GAMBIA (THE)', 90)),
 ('IRAN', ('IRAN (ISLAMIC REPUBLIC OF)', 90)),
 ('ISRAEL', ('UNITED ARAB EMIRATES (THE)', 57)),
 ('MOLDOVA', ('MOLDOVA (THE REPUBLIC OF)', 90)),
 ('MONACO', ('SAMOA', 55)),
 ('NETHERLANDS', ('NETHERLANDS (THE)', 95)),
 ('NIGER', ('NIGER (THE)', 90)),
 ('NORTH KOREA', ('KOREA (THE REPUBLIC OF)', 86)),
 ('PHILIPPINES', ('PHILIPPINES (THE)', 95)),
 ('PUERTO RICO', ('BOLIVIA (PLURINATIONAL STATE OF)', 50)),
 ('RUSSIA', ('RUSSIAN FEDERATION (

In [275]:
changesDf2_1={f:fz.extractOne(f, OnlyISOS)[0] 
                 for f in sorted(OnlyDf2)
                 if fz.extractOne(f, OnlyISOS)[1] >=87}
#dict of matches
changesDf2_1

{'BOLIVIA': 'BOLIVIA (PLURINATIONAL STATE OF)',
 'BRUNEI': 'BRUNEI DARUSSALAM',
 'CENTRAL AFRICAN REPUBLIC': 'CENTRAL AFRICAN REPUBLIC (THE)',
 'COMOROS': 'COMOROS (THE)',
 'CONGO': 'CONGO (THE)',
 "COTE D'IVOIRE": "CÔTE D'IVOIRE",
 'DEMOCRATIC REPUBLIC OF CONGO': 'CONGO (THE DEMOCRATIC REPUBLIC OF THE)',
 'DOMINICA': 'DOMINICAN REPUBLIC (THE)',
 'DOMINICAN REPUBLIC': 'DOMINICAN REPUBLIC (THE)',
 'GAMBIA': 'GAMBIA (THE)',
 'IRAN': 'IRAN (ISLAMIC REPUBLIC OF)',
 'MOLDOVA': 'MOLDOVA (THE REPUBLIC OF)',
 'NETHERLANDS': 'NETHERLANDS (THE)',
 'NIGER': 'NIGER (THE)',
 'PHILIPPINES': 'PHILIPPINES (THE)',
 'RUSSIA': 'RUSSIAN FEDERATION (THE)',
 'SUDAN': 'SUDAN (THE)',
 'TANZANIA': 'TANZANIA, THE UNITED REPUBLIC OF',
 'UNITED ARAB EMIRATES': 'UNITED ARAB EMIRATES (THE)',
 'UNITED KINGDOM': 'UNITED KINGDOM OF GREAT BRITAIN AND NORTHERN IRELAND (THE)',
 'UNITED STATES': 'UNITED STATES MINOR OUTLYING ISLANDS (THE)',
 'VIETNAM': 'VIET NAM'}

In [278]:
df2.Country.replace(to_replace=changesDf2_1, inplace=True)
df2

COU                                 Country  SMALL_SUBNATIONAL_REGION  \
0    AUS                               AUSTRALIA                       NaN   
1    AUT                                 AUSTRIA                       NaN   
2    BEL                                 BELGIUM                       NaN   
3    CAN                                  CANADA                       NaN   
4    CZE  CONGO (THE DEMOCRATIC REPUBLIC OF THE)                       NaN   
..   ...                                     ...                       ...   
173  ZMB                                  ZAMBIA                       NaN   
174  ZWE                                ZIMBABWE                       NaN   
175  SRB                                  SERBIA                       NaN   
176  MNE                              MONTENEGRO                       NaN   
177  SSD                             SOUTH SUDAN                       NaN   

    Small subnational region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
173           Not applicable                    TOTAL   
174           Not applicable                    TOTAL   
175           Not applicable                    TOTAL   
176           Not applicable                    TOTAL   
177           Not applicable                    TOTAL   

    Large subnational region     VAR                           Variable  \
0                      Total  PWM_EX  Mean population exposure to PM2.5   
1                      Total  PWM_EX  Mean population exposure to PM2.5   
2                      Total  PWM_EX  Mean population exposure to PM2.5   
3                      Total  PWM_EX  Mean population exposure to PM2.5   
4                      Total  PWM_EX  Mean population exposure to PM2.5   
..                       ...     ...                                ...   
173                    Total  PWM_EX  Mean population exposure to PM2.5   
174                    Total  PWM_EX  Mean population exposure to PM2.5   
175                    Total  PWM_EX  Mean population exposure to PM2.5   
176                    Total  PWM_EX  Mean population exposure to PM2.5   
177                    Total  PWM_EX  Mean population exposure to PM2.5   

    Unit Code                        Unit  ...  Reference Period   Value  \
0    MICRO_M3  Micrograms per cubic metre  ...               NaN   7.730   
1    MICRO_M3  Micrograms per cubic metre  ...               NaN  12.720   
2    MICRO_M3  Micrograms per cubic metre  ...               NaN  12.540   
3    MICRO_M3  Micrograms per cubic metre  ...               NaN   7.090   
4    MICRO_M3  Micrograms per cubic metre  ...               NaN  17.560   
..        ...                         ...  ...               ...     ...   
173  MICRO_M3  Micrograms per cubic metre  ...               NaN  23.360   
174  MICRO_M3  Micrograms per cubic metre  ...               NaN  19.160   
175  MICRO_M3  Micrograms per cubic metre  ...               NaN  26.073   
176  MICRO_M3  Micrograms per cubic metre  ...               NaN  23.260   
177  MICRO_M3  Micrograms per cubic metre  ...               NaN  38.849   

     Flag Codes            Flags  Rank Population Date_of_Information  \
0             E  Estimated value    54   26461166           2023 est.   
1             E  Estimated value    99    8940860           2023 est.   
2             E  Estimated value    82   11913633           2023 est.   
3             E  Estimated value    38   38516736           2023 est.   
4             E  Estimated value    14  111859928           2023 est.   
..          ...              ...   ...        ...                 ...   
173           E  Estimated value    63   20216029         

In [280]:
final_df=country_isos.merge(df2, left_on='Country', right_on='Country')
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  COU  SMALL_SUBNATIONAL_REGION  \
0             .af   AF  AFG  AFG                       NaN   
1             .al   AL  ALB  ALB                       NaN   
2             .dz   DZ  DZA  DZA                       NaN   
3             .ao   AO  AGO  AGO                       NaN   
4             .ag   AG  ATG  ATG                       NaN   
..            ...  ...  ...  ...                       ...   
159           .uz   UZ  UZB  UZB                       NaN   
160           .vn   VN  VNM  VNM                       NaN   
161           .ye   YE  YEM  YEM                       NaN   
162           .zm   ZM  ZMB  ZMB                       NaN   
163           .zw   ZW  ZWE  ZWE                       NaN   

    Small subnational region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
159           Not applicable                    TOTAL   
160           Not applicable                    TOTAL   
161           Not applicable                    TOTAL   
162           Not applicable                    TOTAL   
163           Not applicable                    TOTAL   

    Large subnational region  ... Reference Period  Value Flag Codes  \
0                      Total  ...              NaN  68.27          E   
1                      Total  ...              NaN  18.90          E   
2                      Total  ...              NaN  24.93          E   
3                      Total  ...              NaN  24.20          E   
4                      Total  ...              NaN  18.99          E   
..                       ...  ...              ...    ...        ...   
159                    Total  ...              NaN  36.29          E   
160                    Total  ...              NaN  20.05          E   
161                    Total  ...              NaN  44.03          E   
162                    Total  ...              NaN  23.36          E   
163                    Total  ...              NaN  19.16          E   

               Flags  Rank Population  Date_of_Information  \
0    Estimated value    37   39232003            2023 est.   
1    Estimated value   137    3101621            2023 est.   
2    Estimated value    34   44758398            2023 est.   
3    Estimated value    41   35981281            2023 est.   
4    Estimated value   194     101489            2023 est.   
..               ...   ...        ...                  ...   
159  Estimated value    48   31360836            2023 est.   
160  Estimated value    16  104799174            2023 est.   
161  Estimated value    47   31565602            2023 est.   
162  Estimated value    63   20216029            2023 est.   
163  Estimated value    73   15418674            2023 est.   

                  Entity  Code  \
0            Afghanistan   AFG   
1                Albania   ALB  

Note: By the fuzzy merge we loose less rows. With the simple merge between df2 and ISO codes we had 140 rows, but with the fuzzy merge we have 164 rows

In [283]:
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  COU  SMALL_SUBNATIONAL_REGION  \
0             .af   AF  AFG  AFG                       NaN   
1             .al   AL  ALB  ALB                       NaN   
2             .dz   DZ  DZA  DZA                       NaN   
3             .ao   AO  AGO  AGO                       NaN   
4             .ag   AG  ATG  ATG                       NaN   
..            ...  ...  ...  ...                       ...   
159           .uz   UZ  UZB  UZB                       NaN   
160           .vn   VN  VNM  VNM                       NaN   
161           .ye   YE  YEM  YEM                       NaN   
162           .zm   ZM  ZMB  ZMB                       NaN   
163           .zw   ZW  ZWE  ZWE                       NaN   

    Small subnational region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
159           Not applicable                    TOTAL   
160           Not applicable                    TOTAL   
161           Not applicable                    TOTAL   
162           Not applicable                    TOTAL   
163           Not applicable                    TOTAL   

    Large subnational region  ... Reference Period Code Reference Period  \
0                      Total  ...                   NaN              NaN   
1                      Total  ...                   NaN              NaN   
2                      Total  ...                   NaN              NaN   
3                      Total  ...                   NaN              NaN   
4                      Total  ...                   NaN              NaN   
..                       ...  ...                   ...              ...   
159                    Total  ...                   NaN              NaN   
160                    Total  ...                   NaN              NaN   
161                    Total  ...                   NaN              NaN   
162                    Total  ...                   NaN              NaN   
163                    Total  ...                   NaN              NaN   

     Value Flag Codes            Flags Rank  Population  Date_of_Information  \
0    68.27          E  Estimated value   37    39232003            2023 est.   
1    18.90          E  Estimated value  137     3101621            2023 est.   
2    24.93          E  Estimated value   34    44758398            2023 est.   
3    24.20          E  Estimated value   41    35981281            2023 est.   
4    18.99          E  Estimated value  194      101489            2023 est.   
..     ...        ...              ...  ...         ...                  ...   
159  36.29          E  Estimated value   48    31360836            2023 est.   
160  20.05          E  Estimated value   16   104799174            2023 est.   
161  44.03          E  Estimated value   47    31565602            2023 est. 

### Cleaning & formatting
Now we will procede with the cleaning and formatting (if needed)

The variables we want to keep are: 
1. Value: this corresponds to the Mean population exposure to PM2.5. Its units are Micrograms per cubic metre. Note that initially we restricted the original "PM2.5" data frame to only the values from 2018. 
2. Population: It is the number of people by country
3. Annual freshwater withdrawals, total (billion cubic meters): Note that initially we restricted the original "Annual freshwater withdrawls" to the values from year 2018. 

Therefore: 
For column "Value", we will rename it to "PM2.5_exposure_2018"
For column "Annual freshwater withdrawals, total (billion cubic meters)", we will rename it to "Annual_H2O_withdrawl_2018"

In [285]:
# We get rid off of any white spaces between the words of the columns' titles
final_df.columns=final_df.columns.str.replace(" ","_")
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  COU  SMALL_SUBNATIONAL_REGION  \
0             .af   AF  AFG  AFG                       NaN   
1             .al   AL  ALB  ALB                       NaN   
2             .dz   DZ  DZA  DZA                       NaN   
3             .ao   AO  AGO  AGO                       NaN   
4             .ag   AG  ATG  ATG                       NaN   
..            ...  ...  ...  ...                       ...   
159           .uz   UZ  UZB  UZB                       NaN   
160           .vn   VN  VNM  VNM                       NaN   
161           .ye   YE  YEM  YEM                       NaN   
162           .zm   ZM  ZMB  ZMB                       NaN   
163           .zw   ZW  ZWE  ZWE                       NaN   

    Small_subnational_region LARGE_SUBNATIONAL_REGION  \
0             Not applicable                    TOTAL   
1             Not applicable                    TOTAL   
2             Not applicable                    TOTAL   
3             Not applicable                    TOTAL   
4             Not applicable                    TOTAL   
..                       ...                      ...   
159           Not applicable                    TOTAL   
160           Not applicable                    TOTAL   
161           Not applicable                    TOTAL   
162           Not applicable                    TOTAL   
163           Not applicable                    TOTAL   

    Large_subnational_region  ... Reference_Period_Code Reference_Period  \
0                      Total  ...                   NaN              NaN   
1                      Total  ...                   NaN              NaN   
2                      Total  ...                   NaN              NaN   
3                      Total  ...                   NaN              NaN   
4                      Total  ...                   NaN              NaN   
..                       ...  ...                   ...              ...   
159                    Total  ...                   NaN              NaN   
160                    Total  ...                   NaN              NaN   
161                    Total  ...                   NaN              NaN   
162                    Total  ...                   NaN              NaN   
163                    Total  ...                   NaN              NaN   

     Value Flag_Codes            Flags Rank  Population  Date_of_Information  \
0    68.27          E  Estimated value   37    39232003            2023 est.   
1    18.90          E  Estimated value  137     3101621            2023 est.   
2    24.93          E  Estimated value   34    44758398            2023 est.   
3    24.20          E  Estimated value   41    35981281            2023 est.   
4    18.99          E  Estimated value  194      101489            2023 est.   
..     ...        ...              ...  ...         ...                  ...   
159  36.29          E  Estimated value   48    31360836            2023 est.   
160  20.05          E  Estimated value   16   104799174            2023 est.   
161  44.03          E  Estimated value   47    31565602            2023 est. 

In [286]:
# We get rid off of the variables that are not of our interest
delete_columns=["COU","SMALL_SUBNATIONAL_REGION","Small_subnational_region","LARGE_SUBNATIONAL_REGION","Large_subnational_region",
               "Reference_Period_Code","Reference_Period","Flag_Codes","Flags","Rank","Code"]
final_df.drop(columns=delete_columns,inplace=True)
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3     VAR                           Variable  \
0             .af   AF  AFG  PWM_EX  Mean population exposure to PM2.5   
1             .al   AL  ALB  PWM_EX  Mean population exposure to PM2.5   
2             .dz   DZ  DZA  PWM_EX  Mean population exposure to PM2.5   
3             .ao   AO  AGO  PWM_EX  Mean population exposure to PM2.5   
4             .ag   AG  ATG  PWM_EX  Mean population exposure to PM2.5   
..            ...  ...  ...     ...                                ...   
159           .uz   UZ  UZB  PWM_EX  Mean population exposure to PM2.5   
160           .vn   VN  VNM  PWM_EX  Mean population exposure to PM2.5   
161           .ye   YE  YEM  PWM_EX  Mean population exposure to PM2.5   
162           .zm   ZM  ZMB  PWM_EX  Mean population exposure to PM2.5   
163           .zw   ZW  ZWE  PWM_EX  Mean population exposure to PM2.5   

    Unit_Code                        Unit  PowerCode_Code PowerCode  Value  \
0    MICRO_M3  Micrograms per cubic metre               0     Units  68.27   
1    MICRO_M3  Micrograms per cubic metre               0     Units  18.90   
2    MICRO_M3  Micrograms per cubic metre               0     Units  24.93   
3    MICRO_M3  Micrograms per cubic metre               0     Units  24.20   
4    MICRO_M3  Micrograms per cubic metre               0     Units  18.99   
..        ...                         ...             ...       ...    ...   
159  MICRO_M3  Micrograms per cubic metre               0     Units  36.29   
160  MICRO_M3  Micrograms per cubic metre               0     Units  20.05   
161  MICRO_M3  Micrograms per cubic metre               0     Units  44.03   
162  MICRO_M3  Micrograms per cubic metre               0     Units  23.36   
163  MICRO_M3  Micrograms per cubic metre               0     Units  19.16   

     Population Date_of_Information  \
0      39232003           2023 est.   
1       3101621           2023 est.   
2      44758398           2023 est.   
3      35981281           2023 est.   
4        101489           2023 est.   
..          ...                 ...   
159    31360836           2023 est.   
160   104799174           2023 est.   
161    31565602           2023 est.   
162    20216029           2023 est.   
163    15418674           2023 est.   

     Annual_freshwater_withdrawals,_total_(billion_cubic_meters)  
0                                         2.028200e+10            
1                                         9.580000e+08            
2                                         9.802000e+09            
3                                         7.057000e+08            
4                                         4.400000e+06            
..                                                 ...            
159                                       5.890400e+10            
160                                       8.186200e+10            
161                                       3.565000e+09            
162                                       1.572000e+09            
163                                       3.771380e+09            

[164 rows x 15 columns]

In [288]:
# We keep deleting the unnecesary columns
bye_columns=["VAR","Variable","Unit_Code","Unit","PowerCode_Code","PowerCode"]
final_df.drop(columns=bye_columns,inplace=True)
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  Value  Population Date_of_Information  \
0             .af   AF  AFG  68.27    39232003           2023 est.   
1             .al   AL  ALB  18.90     3101621           2023 est.   
2             .dz   DZ  DZA  24.93    44758398           2023 est.   
3             .ao   AO  AGO  24.20    35981281           2023 est.   
4             .ag   AG  ATG  18.99      101489           2023 est.   
..            ...  ...  ...    ...         ...                 ...   
159           .uz   UZ  UZB  36.29    31360836           2023 est.   
160           .vn   VN  VNM  20.05   104799174           2023 est.   
161           .ye   YE  YEM  44.03    31565602           2023 est.   
162           .zm   ZM  ZMB  23.36    20216029           2023 est.   
163           .zw   ZW  ZWE  19.16    15418674           2023 est.   

     Annual_freshwater_withdrawals,_total_(billion_cubic_meters)  
0                                         2.028200e+10            
1                                         9.580000e+08            
2                                         9.802000e+09            
3                                         7.057000e+08            
4                                         4.400000e+06            
..                                                 ...            
159                                       5.890400e+10            
160                                       8.186200e+10            
161                                       3.565000e+09            
162                                       1.572000e+09            
163                                       3.771380e+09            

[164 rows x 9 columns]

In [293]:
# We rename certain column titles
final_df.rename(columns={'Value':'PM2.5_exposure_2018', 'Date_of_Information':'Year_of_population_value',
                        'Annual_freshwater_withdrawals,_total_(billion_cubic_meters)':'Annual_H2O_withdrawl_2018'},inplace=True)
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  PM2.5_exposure_2018  Population  \
0             .af   AF  AFG                68.27    39232003   
1             .al   AL  ALB                18.90     3101621   
2             .dz   DZ  DZA                24.93    44758398   
3             .ao   AO  AGO                24.20    35981281   
4             .ag   AG  ATG                18.99      101489   
..            ...  ...  ...                  ...         ...   
159           .uz   UZ  UZB                36.29    31360836   
160           .vn   VN  VNM                20.05   104799174   
161           .ye   YE  YEM                44.03    31565602   
162           .zm   ZM  ZMB                23.36    20216029   
163           .zw   ZW  ZWE                19.16    15418674   

    Year_of_population_value  Annual_H2O_withdrawl_2018  
0                  2023 est.               2.028200e+10  
1                  2023 est.               9.580000e+08  
2                  2023 est.               9.802000e+09  
3                  2023 est.               7.057000e+08  
4                  2023 est.               4.400000e+06  
..                       ...                        ...  
159                2023 est.               5.890400e+10  
160                2023 est.               8.186200e+10  
161                2023 est.               3.565000e+09  
162                2023 est.               1.572000e+09  
163                2023 est.               3.771380e+09  

[164 rows x 9 columns]

In [305]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 164 entries, 0 to 163
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Country                    164 non-null    object 
 1   Officialstatename          164 non-null    object 
 2   InternetccTLD              163 non-null    object 
 3   iso2                       163 non-null    object 
 4   iso3                       164 non-null    object 
 5   PM2.5_exposure_2018        164 non-null    float64
 6   Population                 164 non-null    int64  
 7   Annual_H2O_withdrawl_2018  164 non-null    float64
dtypes: float64(2), int64(1), object(5)
memory usage: 10.4+ KB


In [304]:
# In column "Year_of_population_value", we need to get rid off est. and white spaces
# Note that python reads the values as objects instead of numbers
final_df['Year_of_population_value']=final_df.Year_of_population_value.str.extract(pat=r'(^\w+|.$)', expand=True)
final_df

AttributeError: 'DataFrame' object has no attribute 'Year_of_population_value'

In [ ]:
#Since all values correspond to year 2023, we can delete that column
final_df=final_df.drop('Year_of_population_value', axis=1)

In [306]:
# We rename the population column name
final_df.rename(columns={'Population':'Population_in_2023','PM2.5_exposure_2018':'PM2_5_exposure_2018'},inplace=True)
final_df

Country                            Officialstatename  \
0            AFGHANISTAN          The Islamic Republic of Afghanistan   
1                ALBANIA                      The Republic of Albania   
2                ALGERIA  The People's Democratic Republic of Algeria   
3                 ANGOLA                       The Republic of Angola   
4    ANTIGUA AND BARBUDA                          Antigua and Barbuda   
..                   ...                                          ...   
159           UZBEKISTAN                   The Republic of Uzbekistan   
160             VIET NAM           The Socialist Republic of Viet Nam   
161                YEMEN                        The Republic of Yemen   
162               ZAMBIA                       The Republic of Zambia   
163             ZIMBABWE                     The Republic of Zimbabwe   

    InternetccTLD iso2 iso3  PM2_5_exposure_2018  Population_in_2023  \
0             .af   AF  AFG                68.27            39232003   
1             .al   AL  ALB                18.90             3101621   
2             .dz   DZ  DZA                24.93            44758398   
3             .ao   AO  AGO                24.20            35981281   
4             .ag   AG  ATG                18.99              101489   
..            ...  ...  ...                  ...                 ...   
159           .uz   UZ  UZB                36.29            31360836   
160           .vn   VN  VNM                20.05           104799174   
161           .ye   YE  YEM                44.03            31565602   
162           .zm   ZM  ZMB                23.36            20216029   
163           .zw   ZW  ZWE                19.16            15418674   

     Annual_H2O_withdrawl_2018  
0                 2.028200e+10  
1                 9.580000e+08  
2                 9.802000e+09  
3                 7.057000e+08  
4                 4.400000e+06  
..                         ...  
159               5.890400e+10  
160               8.186200e+10  
161               3.565000e+09  
162               1.572000e+09  
163               3.771380e+09  

[164 rows x 8 columns]

In [307]:
# Lastly we corroborate that the columns correspond to the variable type (string or object, float and int)
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 164 entries, 0 to 163
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Country                    164 non-null    object 
 1   Officialstatename          164 non-null    object 
 2   InternetccTLD              163 non-null    object 
 3   iso2                       163 non-null    object 
 4   iso3                       164 non-null    object 
 5   PM2_5_exposure_2018        164 non-null    float64
 6   Population_in_2023         164 non-null    int64  
 7   Annual_H2O_withdrawl_2018  164 non-null    float64
dtypes: float64(2), int64(1), object(5)
memory usage: 10.4+ KB
